feat(iota-genesis-builder): Allocate TimeLocks directly into a Validator stake pool in the Genesis

miker83z opened the pull request:
# Description of change

Provides the ability to create a genesis blob with the allocation of timelock objects to validator staking pools.

## Links to any relevant issues

Closes #441 

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

`cargo run --release --example build_stardust_genesis`

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

kodemartin committed:
feat(iota-genesis-builder): bootstrap genesis-stake logic (#706)

kodemartin committed:
feat(iota-genesis-builder): implement MigrationObjects query methods (#714)

* feat(iota-genesis-builder): implement MigrationObjects query methods

* refactor(iota-genesis-builder): optimize MigrationObjects queries

kodemartin committed:
fix(iota-config): [de]serialize Genesis directly (#865)

miker83z committed:
feat(iota-genesis-builder): Add delegate genesis stake logic (#741)

* feat(iota-genesis-builder): add delegate genesis stake logic

* refactor(iota-genesis-builder): add stake module

* feat(iota-genesis-builder): burn staked gas coins

* fixup! refactor(iota-genesis-builder): add stake module

* fixup! fixup! refactor(iota-genesis-builder): add stake module

* refactor(iota-genesis-builder): fix error handling in delegate_genesis_stake

* refactor(iota-genesis-builder): pass target stake as argument

* refactor(iota-genesis-builder): modify TimelockAllocation

* feat(iota-genesis-builder): add timelocked stake to genesis

* refactor(iota-genesis-builder): add license header in stake.rs

* fixup! feat(iota-genesis-builder): add timelocked stake to genesis

* fix(iota-genesis-builder): enable validation of token distribution schedule

* fix(iota-genesis-builder): remove option

* fix(iota-genesis-builder): update timelock allocation PTB

* feat(iota-framework/timelock): Enhance the Public API of `timelock` package (changed title) (#812)

* feat(iota-framework/timelock): create timelock self transfer

* fix(iota-framework/timelock): typos and small refactor

* feat(iota-framework/timelock): add `lock_to` functions

The ability to lock an object to a specific address was missing from the pubic API

---------

Co-authored-by: Levente Pap <levente.pap@iota.org>

* chore(iota-framework-snapshot): temporary update v1 framework snapshots

* fix(iota-genesis-builder): print genesis objects len in example

* fix(iota-genesis-builder): clippy

* fix(iota-genesis-builder): move genesis stake creation back into

* fix(iota-genesis-builder): comments fixes

* fix(iota-framework-snapshot): update v1 snapshot bin

* refactor(iota-framework/timelock): update naming

* refactor(iota-framework/timelock): implement comments

* fix(iota-framework-snapshot): update v1 snapshot bin

* feat(iota-genesis-builder): save genesis blob file

* refactor(iota-framework/timelock): update naming

* fix(iota-framework-snapshot): update v1 snapshot bin

* feat(iota-genesis-builder): sort timelock objects for delegation (#925)

---------

Co-authored-by: Konstantinos Demartinos <konstantinos.demartinos@iota.org>
Co-authored-by: Levente Pap <levente.pap@iota.org>

kodemartin committed:
refactor(iota-genesis-builder): optimize memory usage while building genesis (#914)

* refactor(iota-genesis-builder): avoid cloning in Builder::build_unsigned_genesis_checkpoint

* refactor(iota-genesis-builder): rename Builder:build_unsigned_genesis_checkpoint

* refactor(iota-genesis-builder): consume genesis objects while building

* refactor(iota-config): load/save genesis using read/write buffers

* refactor(iota-genesis-builder): optimize Builder::{load,save}

* refactor(iota-genesis-builder): use expect instead of unwrap

Co-authored-by: Philipp Gackstatter <philipp.gackstatter@iota.org>

* refactor(iota-genesis-builder): remove object.to_owned and satisfy clippy

* fixup! refactor(iota-genesis-builder): rename Builder:build_unsigned_genesis_checkpoint

---------

Co-authored-by: Philipp Gackstatter <philipp.gackstatter@iota.org>

miker83z committed:
fix(iota-genesis-builder): Timelock allocations execution (#918)

* fix(iota-genesis-builder): genesis stake timelock_allocations memory leak

* fix(iota-genesis-builder): timelock allocation execution with inputs

* refactor(iota-genesis-builder): read from uncompressed snapshot

* feat(iota-genesis-builder): add vanilla build example

* fix(iota-genesis-builder): make timelock allocation PTB optional

* refactor(iota-genesis-builder): update examples with token schedule

* fix(iota-genesis-builder): use IotaSystemStateV2 in validate_output

* fix(iota-genesis-builder): reintroduce public(package) transfer in Timelock modules

* fix(iota-framework-snapshot): update v1 snapshot bin

miker83z committed:
refactor(iota-framework/timelock): Split `timelock` package into framework and system (#1030)

* refactor(iota-framework/timelock): split timelock package into framework and system

* feat(iota-system): create a timelocked request_add_stake_at_genesis

* chore(iota-framework): remove timelock package and update docs

* fix(iota-framework-snapshot): update v1 snapshot bin

* fix(crates): update timelock addresses and make all crates build

* fix(iota-framework-tests): remove timelock tests

* refactor(iota-system): change internal stake fun name

* fix(iota-framework): fun visibility and label type name

* fix(iota-framework): packages builds

* fix(iota-framework-snapshot): remove dirty

miker83z committed:
Merge branch 'develop' into sc-platform/genesis-staking

miker83z committed:
refactor(iota-genesis-builder): Adapt `TokenAllocation` to support timelocks (#1105)

* refactor(iota-config/genesis): add expiration timestamps to TokenAllocation

* fix(iota-genesis-builder): adapt builder to new TokenAllocation

* fix(iota-genesis-builder): clippy

* fix(docs): remove timelock package

* refactor(iota-config/genesis): make TokenAllocation more granular for timelocks

* fix(iota-genesis-builder): loop in timelock delegation

* fix(iota-genesis-builder): fixes after comments

---------

Co-authored-by: Levente Pap <levente.pap@iota.org>

miker83z committed:
fix(iota-system): Adapt `genesis::create` to new changes of `TokenAllocation` (#1130)

* fix(iota-system): adapt to new token allocation

* fix(iota-genesis-builder): add inputs to new genesis create fun

* fix(iota-framework-snapshot): remove dirty

* fix(iota-framework-snapshot): remove dirty

miker83z committed:
fix(iota-framework-snapshot): update v1 snapshot manifest

miker83z committed:
Merge branch 'develop' into sc-platform/genesis-staking

samuel-rufi reviewed (dismissed):
Hey @miker83z - great PR!

lzpap committed:
fix(iota-genesis-builder): Enable SystemStateInnerV1 in genesis

https://github.com/iotaledger/iota/issues/1083 will do the cleanup of
outdated system object versions, until then this check makes
tests fail.

miker83z committed:
fix(iota-genesis-builder): validation of timelock stake iota

miker83z committed:
Merge branch 'develop' into sc-platform/genesis-staking

miker83z committed:
fix(iota-genesis-builder): CI tests
